python/cudnn/linear_attention/cutile/kda_engine.py

```
@@ -289,12 +208,100 @@ def _execute_bwd(self, node, nb, bufs, stream) -> None:
+        f32 = cudnn.data_type.FLOAT
+        for port, got in (
+            ("initial_state", facts.state_dtype),
+            ("final_state", facts.final_state_dtype),
+            ("d_final_state", facts.d_final_state_dtype),
+            ("d_initial_state", facts.d_initial_state_dtype),
+            ("a_log", facts.a_log_dtype),
+            ("dt_bias", facts.dt_bias_dtype),
+        ):
+            if got not in (f32, None):
+                raise NotImplementedError(f"KdaCuTileEngine: '{port}' must be fp32 (callers convert), got {got}")
+        glb = graph.nodes and list(graph.nodes)[0].params.get("gate_lower_bound")
```

Ruff (0.16.1)

[warning] Prefer `next(iter(graph.nodes))` over single element slice

Replace with `next(iter(graph.nodes))`

(RUF015)